File: frontend/src/types/consultantAssignment.ts
import { User } from './index'
import { Project } from './index'

export type AssignmentStatus = 'pending' | 'active' | 'completed' | 'cancelled'

export interface ConsultantType {
  id: string
  name: string
  nameHe: string
  category?: string
  description?: string
  isActive: boolean
  createdAt: string
  updatedAt: string
}

export interface ConsultantAssignment {
  id: string
  consultantId: string
  projectId: string
  consultantTypeId?: string
  startDate: string
  endDate: string
  status: AssignmentStatus
  notes?: string
  createdAt: string
  updatedAt: string
  consultant?: User
  project?: Project
  consultantType?: ConsultantType
}

export interface ConsultantAssignmentCreate {
  consultantId: string
  projectId: string
  consultantTypeId?: string
  startDate: string
  endDate: string
  status?: AssignmentStatus
  notes?: string
}

export interface ConsultantAssignmentUpdate {
  consultantId?: string
  projectId?: string
  consultantTypeId?: string
  startDate?: string
  endDate?: string
  status?: AssignmentStatus
  notes?: string
}

File: backend/app/models/consultant_assignment.py
from __future__ import annotations

import uuid
from datetime import date, datetime, timezone
from enum import Enum
from typing import Optional

from sqlalchemy import Date, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class AssignmentStatus(str, Enum):
    PENDING = "pending"
    ACTIVE = "active"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ConsultantAssignment(Base):
    __tablename__ = "consultant_assignments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    consultant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"))
    project_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"))
    consultant_type_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), ForeignKey("consultant_types.id", ondelete="SET NULL"))
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(String(50), default=AssignmentStatus.PENDING.value)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    consultant = relationship("User", foreign_keys=[consultant_id])
    project = relationship("Project", foreign_keys=[project_id])
    consultant_type = relationship("ConsultantType", foreign_keys=[consultant_type_id])
